File: src/nlp/entity_recognizer.py
from dataclasses import dataclass
from typing import List, Dict, Pattern, Optional
import re
from .tokenizer import Token
from .data.symptoms import SymptomData, symptoms_database

@dataclass
class Entity:
    """Represents a recognized entity from the text."""
    type: str  # 'symptom' | 'severity' | 'duration' | 'bodyPart' | 'medication'
    value: str
    confidence: float
    start_position: int
    end_position: int

class EntityRecognizer:
    """Recognizes various medical entities from tokenized text."""

    SEVERITY_INDICATORS: Dict[str, List[str]] = {
        'mild': ['mild', 'slight', 'minor', 'little'],
        'moderate': ['moderate', 'medium', 'uncomfortable'],
        'severe': ['severe', 'intense', 'extreme', 'worst', 'unbearable']
    }

    DURATION_PATTERNS: Dict[str, Pattern] = {
        'specific': re.compile(r'(\d+)\s*(day|week|month|year)s?', re.I),
        'relative': re.compile(r'(few|couple|several)\s*(day|week|month|year)s?', re.I),
        'timeOfDay': re.compile(r'(morning|afternoon|evening|night)', re.I)
    }

    BODY_PARTS: List[str] = [
        'head', 'chest', 'stomach', 'back', 'arm', 'leg', 'throat',
        'neck', 'shoulder', 'knee', 'ankle', 'wrist', 'elbow'
    ]

    def recognize_entities(self, tokens: List[Token]) -> List[Entity]:
        """
        Recognize entities from tokenized text.
        
        Args:
            tokens: List of Token objects
            
        Returns:
            List of recognized Entity objects
        """
        text = ' '.join(t.value for t in tokens)
        entities: List[Entity] = []

        entities.extend(self._recognize_symptoms(text))
        entities.extend(self._recognize_severity(text))
        entities.extend(self._recognize_duration(text))
        entities.extend(self._recognize_body_parts(text))

        return entities

    def _recognize_symptoms(self, text: str) -> List[Entity]:
        entities: List[Entity] = []
        
        for symptom in symptoms_database:
            for keyword in symptom.keywords:
                for match in re.finditer(rf'\b{keyword}\b', text, re.I):
                    entities.append(Entity(
                        type='symptom',
                        value=symptom.name,
                        confidence=self._calculate_symptom_confidence(text, symptom),
                        start_position=match.start(),
                        end_position=match.end()
                    ))

        return entities

    def _recognize_severity(self, text: str) -> List[Entity]:
        entities: List[Entity] = []

        for severity, indicators in self.SEVERITY_INDICATORS.items():
            for indicator in indicators:
                for match in re.finditer(rf'\b{indicator}\b', text, re.I):
                    entities.append(Entity(
                        type='severity',
                        value=severity,
                        confidence=0.8,
                        start_position=match.start(),
                        end_position=match.end()
                    ))

        return entities

    def _recognize_duration(self, text: str) -> List[Entity]:
        entities: List[Entity] = []

        for pattern_type, pattern in self.DURATION_PATTERNS.items():
            for match in pattern.finditer(text):
                entities.append(Entity(
                    type='duration',
                    value=match.group(),
                    confidence=0.9,
                    start_position=match.start(),
                    end_position=match.end()
                ))

        return entities

    def _recognize_body_parts(self, text: str) -> List[Entity]:
        entities: List[Entity] = []

        for body_part in self.BODY_PARTS:
            for match in re.finditer(rf'\b{body_part}\b', text, re.I):
                entities.append(Entity(
                    type='bodyPart',
                    value=body_part,
                    confidence=0.9,
                    start_position=match.start(),
                    end_position=match.end()
                ))

        return entities

    def _calculate_symptom_confidence(self, text: str, symptom: SymptomData) -> float:
        confidence = 0.0
        text_lower = text.lower()
        
        # Exact keyword matches
        exact_matches = sum(1 for keyword in symptom.keywords 
                          if keyword.lower() in text_lower)
        confidence += exact_matches * 0.3

        # Related symptom matches
        related_matches = sum(1 for related in symptom.related_symptoms 
                            if related.lower() in text_lower)
        confidence += related_matches * 0.2

        # Context boost
        if 'doctor' in text_lower or 'hospital' in text_lower:
            confidence += 0.1

        return min(confidence, 1.0)

File: src/lib/nlp/entityRecognizer.ts
import { Token } from './tokenizer';
import { symptomsDatabase, SymptomData } from '../../data/symptoms';

export interface Entity {
  type: 'symptom' | 'severity' | 'duration' | 'bodyPart' | 'medication';
  value: string;
  confidence: number;
  startPosition: number;
  endPosition: number;
}

export class EntityRecognizer {
  private static readonly SEVERITY_INDICATORS = {
    mild: ['mild', 'slight', 'minor', 'little'],
    moderate: ['moderate', 'medium', 'uncomfortable'],
    severe: ['severe', 'intense', 'extreme', 'worst', 'unbearable']
  };

  private static readonly DURATION_PATTERNS = {
    specific: /(\d+)\s*(day|week|month|year)s?/i,
    relative: /(few|couple|several)\s*(day|week|month|year)s?/i,
    timeOfDay: /(morning|afternoon|evening|night)/i
  };

  private static readonly BODY_PARTS = [
    'head', 'chest', 'stomach', 'back', 'arm', 'leg', 'throat',
    'neck', 'shoulder', 'knee', 'ankle', 'wrist', 'elbow'
  ];

  recognizeEntities(tokens: Token[]): Entity[] {
    const entities: Entity[] = [];
    const text = tokens.map(t => t.value).join(' ');

    // Recognize symptoms
    entities.push(...this.recognizeSymptoms(text, tokens));

    // Recognize severity
    entities.push(...this.recognizeSeverity(text, tokens));

    // Recognize duration
    entities.push(...this.recognizeDuration(text));

    // Recognize body parts
    entities.push(...this.recognizeBodyParts(text, tokens));

    return entities;
  }

  private recognizeSymptoms(text: string, tokens: Token[]): Entity[] {
    const entities: Entity[] = [];
    
    symptomsDatabase.forEach(symptom => {
      symptom.keywords.forEach(keyword => {
        const regex = new RegExp(`\\b${keyword}\\b`, 'gi');
        let match;
        
        while ((match = regex.exec(text)) !== null) {
          entities.push({
            type: 'symptom',
            value: symptom.name,
            confidence: this.calculateSymptomConfidence(text, symptom),
            startPosition: match.index,
            endPosition: match.index + match[0].length
          });
        }
      });
    });

    return entities;
  }

  private recognizeSeverity(text: string, tokens: Token[]): Entity[] {
    const entities: Entity[] = [];

    Object.entries(EntityRecognizer.SEVERITY_INDICATORS).forEach(([severity, indicators]) => {
      indicators.forEach(indicator => {
        const regex = new RegExp(`\\b${indicator}\\b`, 'gi');
        let match;

        while ((match = regex.exec(text)) !== null) {
          entities.push({
            type: 'severity',
            value: severity,
            confidence: 0.8,
            startPosition: match.index,
            endPosition: match.index + match[0].length
          });
        }
      });
    });

    return entities;
  }

  private recognizeDuration(text: string): Entity[] {
    const entities: Entity[] = [];

    Object.entries(EntityRecognizer.DURATION_PATTERNS).forEach(([type, pattern]) => {
      const matches = text.matchAll(pattern as RegExp);
      for (const match of matches) {
        if (match.index !== undefined) {
          entities.push({
            type: 'duration',
            value: match[0],
            confidence: 0.9,
            startPosition: match.index,
            endPosition: match.index + match[0].length
          });
        }
      }
    });

    return entities;
  }

  private recognizeBodyParts(text: string, tokens: Token[]): Entity[] {
    const entities: Entity[] = [];

    EntityRecognizer.BODY_PARTS.forEach(bodyPart => {
      const regex = new RegExp(`\\b${bodyPart}\\b`, 'gi');
      let match;

      while ((match = regex.exec(text)) !== null) {
        entities.push({
          type: 'bodyPart',
          value: bodyPart,
          confidence: 0.9,
          startPosition: match.index,
          endPosition: match.index + match[0].length
        });
      }
    });

    return entities;
  }

  private calculateSymptomConfidence(text: string, symptom: SymptomData): number {
    let confidence = 0;
    
    // Check for exact keyword matches
    const exactMatches = symptom.keywords.filter(keyword => 
      text.toLowerCase().includes(keyword.toLowerCase())
    ).length;
    confidence += exactMatches * 0.3;

    // Check for related symptoms
    const relatedMatches = symptom.relatedSymptoms.filter(related =>
      text.toLowerCase().includes(related.toLowerCase())
    ).length;
    confidence += relatedMatches * 0.2;

    // Context-based confidence boost
    if (text.toLowerCase().includes('doctor') || 
        text.toLowerCase().includes('hospital')) {
      confidence += 0.1;
    }

    return Math.min(confidence, 1);
  }
}